src/sql/src/statements/option_map.rs

```diff
@@ -12,52 +12,144 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-mod visit;
-mod visit_mut;
-
-use std::borrow::Borrow;
 use std::collections::HashMap;
-use std::iter::FromIterator;
+use std::ops::ControlFlow;
+
+use itertools::Itertools;
+use secrecy::{ExposeSecret, SecretString};
+use sqlparser::ast::{Visit, VisitMut, Visitor, VisitorMut};
+
+const REDACTED_OPTIONS: [&str; 2] = ["access_key_id", "secret_access_key"];
 
 /// Options hashmap.
-/// Because the trait `Visit` and `VisitMut` is not implemented for `HashMap<String, String>`, we have to wrap it and implement them by ourself.
-#[derive(Clone, Eq, PartialEq, Debug)]
+#[derive(Clone, Debug, Default)]
 pub struct OptionMap {
-    pub map: HashMap<String, String>,
+    options: HashMap<String, String>,
+    secrets: HashMap<String, SecretString>,
 }
 
 impl OptionMap {
     pub fn insert(&mut self, k: String, v: String) {
-        self.map.insert(k, v);
+        if REDACTED_OPTIONS.contains(&k.as_str()) {
+            self.secrets.insert(k, SecretString::new(v));
+        } else {
+            self.options.insert(k, v);
+        }
     }
 
     pub fn get(&self, k: &str) -> Option<&String> {
-        self.map.get(k)
+        if let Some(value) = self.options.get(k) {
+            Some(value)
+        } else if let Some(value) = self.secrets.get(k) {
+            Some(value.expose_secret())
+        } else {
+            None
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.options.is_empty() && self.secrets.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.options.len() + self.secrets.len()
+    }
+
+    pub fn to_str_map(&self) -> HashMap<&str, &str> {
+        let mut map = HashMap::new();
+        map.extend(self.options.iter().map(|(k, v)| (k.as_str(), v.as_str())));
+        map.extend(
+            self.secrets
+                .iter()
+                .map(|(k, v)| (k.as_str(), v.expose_secret().as_str())),
+        );
+        map
+    }
+
+    pub fn into_map(self) -> HashMap<String, String> {
+        let mut map = self.options;
+        map.extend(
+            self.secrets
+                .into_iter()
+                .map(|(k, v)| (k, v.expose_secret().to_string())),
```

Comment:
Extra clone in `to_string`. The `secrecy` upstream doesn't export a method for something like `into_string`.

They're experiencing a rewrite so I'm unsure if they can even accept patches:

https://github.com/iqlusioninc/crates/tree/main/secrecy

Reply:
This is unavoidable because the SecretBox implements `Drop` in a customized way. I suppose it won't introduce too much overhead.